### ISSUE TITLE
Issue in Lightly Gaussian Blur Augmentation

### ISSUE THREAD

**Issue by AurelienGauffre** [lightly-ai/lightly#1051]
Hi,
Thank you for the amazing work on this repo. I have identified what might be an issue with the implementation of Lightly's Gaussian Blur that is negatively impacting the performance of most of the Lightly models. The source of the problem appears to be a confusion around the interpretation of the "radius" parameter in PIL's Gaussian Blur filter. Simply removing the blur augmentation currently increases significantly performances on a few Lightly models I have tested. With a quick fix of this blur augmentation, all the models I have tested seem to perform even better (cf. table below), except for SwaV since I have noticed that blur augmentation [has been disabled](https://github.com/lightly-ai/lightly/blob/5f7613b73c59a0418b449f0d7c3e1c7e2262471c/lightly/data/collate.py#L354). This could also explain why SwaV outperforms some models by far in current [Lightly Imagenette Benchmark](https://docs.lightly.ai/self-supervised-learning/getting_started/benchmarks.html) .

## 1. What should be done according to SimClr

According to its documentation, the [ImageCollateFunction()](https://github.com/lightly-ai/lightly/blob/5f7613b73c59a0418b449f0d7c3e1c7e2262471c/lightly/data/collate.py#L90) class from Lightly is designed to replicate the augmentation techniques outlined in the original [SimCLR](https://arxiv.org/abs/2002.05709) paper. Concerning the Gaussian Blur augmentation parameters, the paper describes in Appendix A. :

"We blur the image 50% of the time using a Gaussian kernel. We randomly sample σ ∈ [0.1, 2.0], and the kernel size is set to be 10% of the image height/width." Indeed, a proper Gaussian Blur filter requires two parameters to be defined: the size of the convolution kernel (or, equivalently, the radius, which is half the size of the kernel) and the standard deviation (σ) of this Gaussian kernel. 

## 2. How it is currently done in Lightly VS elsewhere

The Gaussian blur PIL filter used in Lightly only allows for the input of one parameter, labeled as "radius". This lead to potential confusion, as it is unclear how this parameter relates to kernel size and σ. This question has already been raised several times including on [Torchvision github](https://github.com/pytorch/vision/issues/5194) or on [Stackoverflow](https://stackoverflow.com/questions/62968174/for-pil-imagefilter-gaussianblur-how-what-kernel-is-used-and-does-the-radius-par), without very clear answers though, since PIL approximates Gaussian Kernel filter with box linear filter.

The official [documentation of PIL's Gaussian Blur](https://github.com/python-pillow/Pillow/blob/76478c6865c78af10bf48868345db2af92f86166/src/PIL/ImageFilter.py#L154) as well as the second thread above suggests that, non-intuitively, σ should be passed as radius parameter of PIL GaussianBlur function, in contrast to what is done currently in Lightly, where the radius is passed as radius. And in facts, all official implementations of libraries like [MoCo](https://github.com/facebookresearch/moco/blob/5a429c00bb6d4efdf511bf31b6f01e064bf929ab/moco/loader.py#L31), [SwAV](https://github.com/facebookresearch/swav/blob/06b1b7cbaf6ba2a792300d79c7299db98b93b7f9/src/multicropdataset.py#L82), [SimSiam](https://github.com/facebookresearch/simsiam/blob/da2add2749854f9ef459911a8264ecb0b802c1b6/simsiam/loader.py#L31) or [VicReg](https://github.com/facebookresearch/vicreg/blob/4e12602fd495af83efd1631fbe82523e6db092e0/augmentations.py#L21) are also using Gaussian blur from PIL and indeed gives **σ as** **radius parameter in the PIL’s** Gaussian Blur filter (and does not use any radius or kernel size parameter).

This results in blurring images with a strength of one or two order of magnitude higher than what is described in SimCLR, making images barely recognizable with raw eye (with a probability .5, which allows the model to still learn).

Moreover, Lightly currently uses a [scale argument](https://github.com/lightly-ai/lightly/blob/93db513650a1dcc5a9d791c99519ed77269ead11/lightly/transforms/gaussian_blur.py#L23) in its Gaussian Blur function as well as a [normal sampling and a min/max clipping](https://github.com/lightly-ai/lightly/blob/93db513650a1dcc5a9d791c99519ed77269ead11/lightly/transforms/gaussian_blur.py#L52), the origin of which are unclear.

## 3. Solution and results :

Using the [Lightly Imagenette benchmark file](https://docs.lightly.ai/self-supervised-learning/getting_started/benchmarks.html) as is, I have run 3 experiments with a batch size of 256 on a few models (with only one seed) :

**a. Baseline :** I have tried to reproduce the results reported in Lightly Imagenette file to make sure my own baseline is correct. Results are close (except for SwaV on 200 epochs which is weird, since on 800 epochs I got the same ~90% accuracy, I think it might be a mistake there)
**b. Ablation without any blur :** I have just removed the blur from augmentations (put probability of applying blur for all models to 0)
**c. Blur function fixed :** I have fixed the blur function to make it work exactly as in MoCo, SwAV, SimSiam and VicReg official implementations, passing σ to PIL's Gaussian Blur function. The code can be found in #1052 

| Model   | Epochs | KNN Acc % <br>Baseline (from Lightly) | (a) KNN Acc %<br>Baseline (reproduction) | (b) KNN Acc %<br>wihtout Blur | (c) KNN Acc %<br>with Blur solved |
| ------- | ------ | -------------------- | ---------------------------------------- | ----------------------------- | --------------------------------- |
| SimCLR  | 200    | 77.1                 | 77.4                                     | 83.1                          | 84.1                              |
| BYOL    | 200    | 61.9                 | 62.9                                     | 70.1                          | 71.2                              |
| MOCO    | 200    | 72.7                 | 72.9                                     | 78.1                          | 78.7                              |
| SimSiam | 200    | 66.9                 | 67.7                                     | 65.1                          | 75.8                              |
| SwaV    | 200    | 74.8                 | 86.4                                     | 85.6                          | 86.0*                            |
| SimCLR  | 800    | 85.8                 | 85.5                                     | NA                            | 89.2                              |

*I forget to remove the probability of blurring in SwaV to 0 as explained in the intro, so this result does not represent a fixed version of SwaV with Blur, which might explain there is no improvement here compared with baseline. I will replace with the proper result when my training is complete.

The observed gain of ~1 accuracy between b. and c. is roughly on par to what is described in SimCLR paper albeit done with smaller architecture and smaller dataset here. (”We find it helpful, as it improves our ResNet-50 trained for 100 epochs from 63.2% to 64.5%.”) 

PS : It appears that SimCLR has implemented its [own custom version](https://github.com/google-research/simclr/blob/2fc637bdd6a723130db91b377ac15151e01e4fc2/data_util.py#L328) of Gaussian Blur filtering. Through my testing, I have found that there are small differences between SimCLR's approach and PIL's approach, since PIL's does not take into account the kernel size. This means that most of the current official implementations of the Gaussian Blur in SSL papers are not perfectly aligned with the one outlined in the SimCLR official paper and implementation, which probably has very a low impact on the final accuracy, but this still may be a question of interest for the way to implement it in Lightly. Moreover, Torchvision has since released [an  implementation](https://pytorch.org/vision/main/generated/torchvision.transforms.GaussianBlur.html) of Gaussian blur (following a [request](https://github.com/pytorch/vision/pull/2658) about from SSL field), that uses both kernel size and standard deviation σ. From what I have tested, the results are extremely close to the implementation of SimCLR's Gaussian Noise.

**Comment by IgorSusmelj** [comment]
Hey @AurelienGauffre, thank you very much for the contribution. Thanks a lot for investigating this so thoroughly and providing a solution. We will look into your PR + run all the benchmarks again :)
